packages/application/src/layoutrestorer.ts

```diff
@@ -35,11 +35,29 @@ export interface ILayoutRestorer extends IRestorer {
    */
   restored: Promise<void>;
 
+  /**
+   * Whether full layout restoration is deferred and is currently incomplete.
+   *
+   * #### Notes
+   * This flag is useful for tracking when the application has started in
+   * 'single-document' mode and the main area has not yet been restored.
+   */
+  isDeferred: boolean;
```

Comment:
Since these new fields are added to the interface (not optional), it looks like downstream extensions / apps will need to handle them now. How about adding them to the migration guide?

https://jupyterlab.readthedocs.io/en/latest/extension/extension_migration.html#jupyterlab-3-x-to-4-x

Reply:
You're right. In order to make a back-port of this feature, can we make it `isDeferred?: boolean;` for the `3.x` version so that clients in that version check `restorer.isDeferred?` instead of `restorer.isDeferred` to opt-in to this functionality?

Reply:
According to the diff should this also apply to `fetch()` and `restoreDeferred()` since they were also added?

Reply:
Hm, I guess we have to stretch pretty far if we want this in 3.x -- it is possible but maybe not the most elegant if we have to add three optional items to layout restorer.

Reply:
For `3.x` we could create an `IDeferrableRestorer` that is `@deprecated` and use that in the places where we want this functionality and leave `ILayoutRestorer` alone and only upgrade it for `4.x` the way we do in this PR?

Reply:
As these functionalities are only used by the shell currently, would it be a bad practice to declare `LabShell._layoutrestorer` as `LayoutRestorer` instead of `ILayoutRestorer` for 3.x (without changing the  interface)
- here https://github.com/jupyterlab/jupyterlab/blob/0082cb4ee538e8f11b39359cd41e163b9eb08ad8/packages/application/src/shell.ts#L1065
- and here https://github.com/jupyterlab/jupyterlab/blob/0082cb4ee538e8f11b39359cd41e163b9eb08ad8/packages/application/src/shell.ts#L1720

and to update the interface for the 4.x.

Reply:
This is a good idea, @brichet. Perhaps this is the right move in general, instead of modifying `ILayoutRestorer`. I'll think some more about this. What do others think?

Reply:
I opened a PR to do this: https://github.com/jupyterlab/jupyterlab/pull/13300

Reply:
Thanks @afshin